rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,199 @@
+use arrow_array::types::{UInt32Type, UInt64Type, UInt8Type};
+use arrow_array::UInt32Array;
+use futures::{stream, StreamExt};
+use kdam::BarExt;
+use lance_core::datatypes::Schema;
+use lance_core::io::{FileReader, FileWriter, ReadBatchParams, RecordBatchStream};
+
+use lance_core::io::object_store::ObjectStore;
+use lance_core::{Error, Result, ROW_ID};
+use lance_index::vector::{PART_ID_COLUMN, PQ_CODE_COLUMN};
+use log::info;
+use object_store::path::Path;
+use snafu::{location, Location};
+use tempfile::TempDir;
+
+const UNSORTED_BUFFER: &str = "unsorted.lance";
+
+fn get_temp_dir() -> Result<Path> {
+    let dir = TempDir::new()?;
+    let tmp_dir_path = Path::from_filesystem_path(dir.path()).map_err(|e| Error::IO {
+        message: format!("failed to get buffer path in shuffler: {}", e),
+        location: location!(),
+    })?;
+    Ok(tmp_dir_path)
+}
+
+pub struct IvfShuffler {
+    pub num_partitions: u32,
+
+    pub pq_width: usize,
+
+    pub output_dir: Path,
+
+    pub schema: Schema,
+}
+
+impl IvfShuffler {
+    pub fn try_new(
+        num_partitions: u32,
+        pq_width: usize,
+        output_dir: Option<Path>,
+        schema: Schema,
+    ) -> Result<Self> {
+        let output_dir = match output_dir {
+            Some(output_dir) => output_dir,
+            None => get_temp_dir()?,
+        };
+
+        Ok(Self {
+            num_partitions,
+            pq_width,
+            output_dir,
+            schema,
+        })
+    }
+
+    pub async fn dump_unsorted_stream(
+        &self,
+        data: impl RecordBatchStream + Unpin + 'static,
+    ) -> Result<()> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let writer = object_store.create(&path).await?;
+
+        let mut file_writer =
+            FileWriter::with_object_writer(writer, self.schema.clone(), &Default::default())?;
+
+        let mut data = Box::pin(data);
+
+        while let Some(batch) = data.next().await {
+            file_writer.write(&[batch?]).await?;
+        }
+
+        file_writer.finish().await?;
+
+        Ok(())
+    }
+
+    pub async fn count_partition_size(&self) -> Result<Vec<u64>> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        let mut partition_sizes = vec![0; self.num_partitions as usize];
+
+        let lance_schema = reader
+            .schema()
+            .project(&[PART_ID_COLUMN])
+            .expect("part id should exist");
+
+        let mut stream = stream::iter(0..reader.num_batches())
+            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, &lance_schema))
+            .buffered(64);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+            let part_ids: &UInt32Array = as_primitive_array(batch.column(0));
```

Comment:
batch.column(0).as_primitive_array().

Can we do `stream.for_each(..)` instead of while loop?

Reply:
I don't think there is a `array.as_primitive_array()` at least not in our arrow version